### PR TITLE
Adds better support for protos without packages 

### PR DESCRIPTION
**Pull request by sergiocampama**
and more warnings on possible improvements

### CODE REVIEW

**Comment by sergiocampama** [comment]
fyi @thomasvl

File: src/google/protobuf/compiler/objectivec/objectivec_helpers.cc [hunk 1]
```diff
@@ -1026,22 +1026,55 @@ bool ValidateObjCClassPrefix(const FileDescriptor* file,
...
+    } else {
+      // ... no other package has declared that prefix.
+      cerr << endl
+           << "protoc:0: warning: File '" << file->name() << "' has no "
+           << "package. Consider adding a new package to the proto and adding '"
+           << "new.package = " << prefix << "' to the expected prefixes file ("
+           << generation_options.expected_prefixes_path << ")." << endl;
+      cerr.flush();
+    }
+  }
```
Comment (thomasvl) [inline src/google/protobuf/compiler/objectivec/objectivec_helpers.cc]: return true?

i.e. - there's no need for the rest of the checks, right?

File: src/google/protobuf/compiler/objectivec/objectivec_helpers.cc [hunk 2]
```diff
@@ -1026,22 +1026,55 @@ bool ValidateObjCClassPrefix(const FileDescriptor* file,
...
-          generation_options.expected_prefixes_path + ").";
-      return false;  // Only report first usage of the prefix.
+      other_package_for_prefix = i->first;
+      break;
     }
   }
 
+  // Check: Warning - If the file does not have a package, check whether
+  // the prefix declared is being used by another package or not.
+  if (package.length() == 0) {
```
Comment (thomasvl) [inline src/google/protobuf/compiler/objectivec/objectivec_helpers.cc]: package.empty()

reads a little better.

File: src/google/protobuf/compiler/objectivec/objectivec_helpers.cc [hunk 3]
```diff
@@ -1026,22 +1026,55 @@ bool ValidateObjCClassPrefix(const FileDescriptor* file,
...
+           << "new.package = " << prefix << "' to the expected prefixes file ("
+           << generation_options.expected_prefixes_path << ")." << endl;
+      cerr.flush();
+    }
+  }
+
+  // Check: Error - Make sure the prefix wasn't expected for a different
+  // package (overlap is allowed, but it has to be listed as an expected
+  // overlap).
+  if (package.length() > 0 && other_package_for_prefix.length() > 0) {
```
Comment (thomasvl) [inline src/google/protobuf/compiler/objectivec/objectivec_helpers.cc]: With the return, this just becomes if (!other_package_for_prefix.empty()) {

File: src/google/protobuf/compiler/objectivec/objectivec_helpers.cc [hunk 4]
```diff
@@ -1065,7 +1098,7 @@ bool ValidateObjCClassPrefix(const FileDescriptor* file,
 
   // Check: Warning - If the given package/prefix pair wasn't expected, issue a
   // warning issue a warning suggesting it gets added to the file.
-  if (!expected_package_prefixes.empty()) {
+  if (package.length() > 0 && !expected_package_prefixes.empty()) {
```
Comment (thomasvl) [inline src/google/protobuf/compiler/objectivec/objectivec_helpers.cc]: And this addition also isn't needed.

File: src/google/protobuf/compiler/objectivec/objectivec_helpers.cc (hunk 1 above)
Comment (sergiocampama) [inline src/google/protobuf/compiler/objectivec/objectivec_helpers.cc]: I still want the validation on the prefix having a reasonable value done below...

File: src/google/protobuf/compiler/objectivec/objectivec_helpers.cc (hunk 2 above)
Comment (sergiocampama) [inline src/google/protobuf/compiler/objectivec/objectivec_helpers.cc]: Done

File: src/google/protobuf/compiler/objectivec/objectivec_helpers.cc (hunk 1 above)
Comment (thomasvl) [inline src/google/protobuf/compiler/objectivec/objectivec_helpers.cc]: Ahh, good point.  Maybe stick a comment here to explain why we don't return then.

File: src/google/protobuf/compiler/objectivec/objectivec_helpers.cc (hunk 3 above)
Comment (sergiocampama) [inline src/google/protobuf/compiler/objectivec/objectivec_helpers.cc]: Done.

**Comment by thomasvl** [comment]
LG.  Will land once travis finishes on the final change list (kill off the runs on the earlier snapshots).

**Pull request merged**